consensus/src/lib.rs

```diff
@@ -1,27 +1,1095 @@
+            return Err(BlockError::Unknown);
+        }
+
+        // MerkleTree root
+        let merkle_tree_root = block.merkle_root();
+        calculate_tx_merkle_root(block.transactions()).map_or(
+            Err(BlockError::Unknown),
+            |merkle_tree| {
+                if merkle_tree_root != merkle_tree {
+                    Err(BlockError::Unknown)
+                } else {
+                    Ok(())
+                }
+            },
+        )?;
+
+        // Witness merkle root
+        let witness_merkle_root = block.witness_merkle_root();
+        calculate_witness_merkle_root(block.transactions()).map_or(
+            Err(BlockError::Unknown),
+            |witness_merkle| {
+                if witness_merkle_root != witness_merkle {
+                    Err(BlockError::Unknown)
+                } else {
+                    Ok(())
+                }
+            },
+        )?;
+
+        match &block.prev_block_id() {
+            Some(block_id) => {
+                let previous_block = self
+                    .db_tx
+                    .get_block_index(&Id::<Block>::new(&block_id.get()))?
+                    .ok_or(BlockError::Orphan)?;
+                // Time
+                let block_time = block.block_time();
+                if previous_block.get_block_time() > block_time {
+                    return Err(BlockError::Unknown);
+                }
+                if i64::from(block_time) > time::get() {
+                    return Err(BlockError::Unknown);
+                }
+            }
+            None => {
+                // This is only for genesis, AND should never come from a peer
+                if block_source != BlockSource::Local {
+                    return Err(BlockError::InvalidBlockSource);
+                };
+            }
+        }
+
+        self.check_transactions(block)?;
+        Ok(())
+    }
+
+    fn check_consensus(&self, block: &Block) -> Result<(), BlockError> {
+        let _consensus_data = block.consensus_data();
+        // TODO: PoW is not in master at the moment =(
+        Ok(())
+    }
+
+    fn check_transactions(&self, block: &Block) -> Result<(), BlockError> {
+        // TODO: Must check for duplicate inputs (see CVE-2018-17144)
+        //      We should discuss - can we add Hash trait to Transaction?
+        //      We will have plenty more checks with inputs\outputs and HashSet\BTreeMap might be more efficient
+        //
+        // let mut keyed = HashSet::new();
+        // for tx in block.get_transactions() {
+        //     for input in tx.get_inputs() {
+        //         if keyed.contains(input.get_outpoint()) {
+        //             return Err(BlockError::Unknown);
+        //         }
+        //         keyed.insert(input.get_outpoint());
+        //     }
+        // }
+
+        //TODO: Size limits
+        if Encode::encoded_size(block) > MAX_BLOCK_WEIGHT {
```

Comment:
If the `Encode` trait is in scope (which seems to be the case), you should be able to call this as `block.encoded_size()`

Reply:
Cool, I fixed!